### PR TITLE
Ch14501/voice clipping detection

### PR DESCRIPTION
**Pull request by karthikbr82**
https://app.clubhouse.io/callstatsio/story/14501/voice-clipping-detection

Based on - https://www.oreilly.com/library/view/web-audio-api/9781449332679/ch03.html

- [x] Clipping Detection Implementation in VAD module
- [x] Sending clipping event to callstats.js
- [ ] Callback notification in js
- [x] End-to-End test

### CODE REVIEW

File: amazon-connect/callstats-amazon-connect-shim.js [hunk 1]
```diff
@@ -19,9 +24,35 @@
...
+        this.isClipping = false;
+        this.callback('ClippingStop');
+      }
+    }
+
+    checkClipping(audioBuffer) {
+      // Iterate through buffer to check if any of the |values| exceeds 1.
+      for (var i = 0; i < audioBuffer.length; i++) {
+        var absValue = Math.abs(audioBuffer[i]);
+        if (absValue >= 1.0) {
```
Comment (marcinag) [inline amazon-connect/callstats-amazon-connect-shim.js:54]: Should it be > 1.0?

**Review by marcinag — APPROVED** [review]
LGTM. Did you manage to get the callback to work? I assume we attach it only on the sender side, right? I'm not sure if it makes sense to measure it on the receiver side.

File: amazon-connect/callstats-amazon-connect-shim.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/*! callstats Amazon SHIM version = 1.2.0 */
+/*! callstats Amazon SHIM version = 1.2.1 */
```
Comment (mvarela) [inline amazon-connect/callstats-amazon-connect-shim.js]: Unrelated to this PR, but why is the word "SHIM" capitalized like that everywhere? It's not an acronym...

File: amazon-connect/callstats-amazon-connect-shim.js [hunk 3]
```diff
@@ -11,6 +11,11 @@
       this.fftBins = new Float32Array(this.analyser.frequencyBinCount);
       this.mediaStreamSource = this.audioContext.createMediaStreamSource(stream);
       this.mediaStreamSource.connect(this.analyser);
+
+      this.audioProcessor = this.audioContext.createScriptProcessor(512);
```
Comment (lennart-csio) [inline amazon-connect/callstats-amazon-connect-shim.js:15]: is this deprecated? https://developer.mozilla.org/en-US/docs/Web/API/ScriptProcessorNode

File: amazon-connect/callstats-amazon-connect-shim.js [hunk 4]
```diff
@@ -19,9 +24,35 @@
...
+        this.callback('ClippingStop');
+      }
+    }
+
+    checkClipping(audioBuffer) {
+      // Iterate through buffer to check if any of the |values| exceeds 1.
+      for (var i = 0; i < audioBuffer.length; i++) {
+        var absValue = Math.abs(audioBuffer[i]);
+        if (absValue >= 1.0) {
+          return true;
```
Comment (lennart-csio) [inline amazon-connect/callstats-amazon-connect-shim.js]: we could add some measure of clipping. this is boolean, but a single clipped value in the buffer might not be noticeable

File: amazon-connect/callstats-amazon-connect-shim.js (hunk 3 above)
Comment (lennart-csio) [inline amazon-connect/callstats-amazon-connect-shim.js:15]: audioProcessor as well as the analyzer have a window of 512 samples. how often are corresponding functions called?
with a sampling rate of 44kHz there are 44000 samples per second, so 512 samples would only show the last 11.6 msecs

**Pull request merged**